File: android/app/src/main/java/com/jxcode/android/ui/theme/Theme.kt
package com.jxcode.android.ui.theme

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

/**
 * An identity colour: the fill, and the ink that stays legible on it.
 *
 * The ramp is hashed, so any given key can land on any dot and every dot has
 * to carry a readable glyph. White does not: it is 1.81:1 on `amber`, 2.61:1
 * on `orange` and 2.90:1 on `green`, all under the 3:1 floor for a non-text
 * graphic, while the other four are 3.34:1 or better and want white. The inks
 * below are that decision, precomputed — see `Palette.swift` in JXCodeCore for
 * the rule they come from.
 */
@Immutable
data class TileColor(val fill: Color, val ink: Color)

/**
 * The visual language, ported from `Theme.swift` / `Palette.swift` in the
 * macOS app so the two read as one design.
 *
 * What carries over, and why:
 *
 * - **The base is warm, not blue-grey.** Every surface carries a little red
 *   and yellow; nothing here is a neutral grey. `#141319` is the page,
 *   `#191919` a card, `#242427` the raised fill.
 * - **One accent, amber `#FEB43B`.** It is used as a *fill* with dark ink on
 *   top, never as coloured text on a light surface. `accent` (readable amber)
 *   and `accentFill` (the button colour) are separate for that reason.
 * - **Borders are low-alpha white**, so a single constant reads as a hint on
 *   every surface instead of as a stroke tuned to one.
 *
 * The terminal keeps its own ANSI ramp (`TerminalPalette`): a terminal is a
 * well, and an ANSI palette is a functional contract with the programs that
 * emit it, not a brand surface.
 */
object Palette {

    // MARK: - Surfaces

    /// The page behind everything. Deepest of the three.
    val surfaceDeepest = Color(0xFF141319)
    /// Cards and panels — the workhorse surface.
    val surface = Color(0xFF191919)
    /// The selected or hovered fill: a raised panel.
    val surfaceElevated = Color(0xFF242427)
    /// A well — code, terminal, anything inset rather than raised.
    val surfaceSunken = Color(0xFF0F0E13)

    // MARK: - Lines

    /// White at 8%, the desktop's dark-mode stop.
    val border = Color(0x14FFFFFF)
    /// White at 15% — the hover/selected edge.
    val borderStrong = Color(0x26FFFFFF)

    // MARK: - Text

    val textPrimary = Color(0xFFEDEBE8)
    val textSecondary = Color(0xFF9C9C9C)
    /// The inactive tier, held to the *graphic* floor rather than the text
    /// floor: this is what unselected icons and idle dots are drawn in.
    val textTertiary = Color(0xFF717176)

    // MARK: - Accent

    val accent = Color(0xFFFEB43B)
    val accentFill = Color(0xFFFEB43B)
    val accentHover = Color(0xFFFFC55E)
    /// Text and glyphs drawn *on* `accentFill`. Amber with white text fails
    /// contrast, which is the whole reason this exists.
    val accentOn = Color(0xFF1F1E1C)

    // MARK: - Status

    val success = Color(0xFF4BAC65)
    val warning = Color(0xFFF5B549)
    val danger = Color(0xFFF45F59)
    val info = Color(0xFF40C8E0)

    // MARK: - Identity dots

    val tileAmber = TileColor(Color(0xFFF5B549), Color(0xFF1F1E1C))
    val tileBlue = TileColor(Color(0xFF538EDE), Color(0xFFFFFFFF))
    val tileGreen = TileColor(Color(0xFF44AB62), Color(0xFF1F1E1C))
    val tileTeal = TileColor(Color(0xFF41969D), Color(0xFFFFFFFF))
    val tilePurple = TileColor(Color(0xFF845FC9), Color(0xFFFFFFFF))
    val tilePink = TileColor(Color(0xFFAA5BD0), Color(0xFFFFFFFF))
    val tileOrange = TileColor(Color(0xFFF97B64), Color(0xFF1F1E1C))

    /**
     * Deterministic accent for a key, so the same workspace always gets the
     * same dot and the eye learns where it lives.
     *
     * FNV-1a rather than `String.hashCode` — the values have to be stable
     * across processes, which `hashCode` is not (and a relaunch reshuffling
     * every dot is the bug this prevents).
     */
    fun tile(forKey: String): TileColor {
        return when (floorMod(stableHash(forKey), 6)) {
            0 -> tilePurple
            1 -> tileOrange
            2 -> tileTeal
            3 -> tilePink
            4 -> tileGreen
            else -> tileAmber
        }
    }

    /// FNV-1a, 64-bit.
    private fun stableHash(value: String): Long {
        var hash = -3750763034362895579L // 0xcbf29ce484222325
        for (byte in value.toByteArray()) {
            hash = hash xor (byte.toLong() and 0xFF)
            hash *= 0x100000001B3L
        }
        return hash
    }

    private fun floorMod(value: Long, modulus: Int): Int {
        val remainder = (value % modulus).toInt()
        return if (remainder < 0) remainder + modulus else remainder
    }
}

/**
 * The terminal's own colours.
 *
 * `background` and `foreground` are a near-neighbour of `Palette.surfaceSunken`
 * — a terminal is a well, so it is dark in both appearances. The ANSI ramp is
 * functional: it is what the programs running inside the pty emit, so it is
 * not re-sampled to match the chrome.
 */
object TerminalPalette {
    val background = Color(0xFF131418)
    val foreground = Color(0xFFF2F5F7)
    val cursor = Color(0xFFFEB43B)

    val black = Color(0xFF1B2029)
    val red = Color(0xFFE06C75)
    val green = Color(0xFF98C379)
    val yellow = Color(0xFFE5C07B)
    val blue = Color(0xFF61AFEF)
    val magenta = Color(0xFFC678DD)
    val cyan = Color(0xFF56B6C2)
    val white = Color(0xFFD5DAE2)

    val brightBlack = Color(0xFF5C6370)
    val brightRed = Color(0xFFFF7B86)
    val brightGreen = Color(0xFFB5E08C)
    val brightYellow = Color(0xFFF2D194)
    val brightBlue = Color(0xFF7FC4FF)
    val brightMagenta = Color(0xFFDDA0EE)
    val brightCyan = Color(0xFF6FD6E2)
    val brightWhite = Color(0xFFFFFFFF)
}

/**
 * The type scale: small, and deliberately so.
 *
 * These are the sizes the macOS app uses (10pt tertiary, 11pt secondary, 13pt
 * card titles, 21pt the dashboard hero) at sp rather than pt. The Material
 * roles are pinned to them below, because a default `bodyLarge` of 16sp is
 * what makes a Compose screen read as a different app from the SwiftUI one.
 */
object Type {
    val micro = 9.sp
    /// Tertiary text, badges, uppercase section headings.
    val caption = 10.sp
    /// Body copy and mono detail lines.
    val body = 11.sp
    /// Buttons and secondary body.
    val bodyStrong = 12.sp
    /// Card and row titles.
    val title = 13.sp
    /// Product name and screen headings.
    val heading = 14.sp
    /// Sheet titles.
    val sheet = 15.sp
    /// The one large number on a screen.
    val hero = 21.sp

    val mono = FontFamily.Monospace

    /// Uppercase section heading: small, tracked out, tertiary.
    val sectionLabel = TextStyle(
        fontFamily = FontFamily.Default,
        fontWeight = FontWeight.SemiBold,
        fontSize = caption,
        letterSpacing = 0.7.sp,
        color = Palette.textTertiary
    )

    val code = TextStyle(fontFamily = mono, fontSize = body, color = Palette.textSecondary)
}

private val JXCodeTypography = Typography(
    displayLarge = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 32.sp),
    displayMedium = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 26.sp),
    displaySmall = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 22.sp),
    headlineLarge = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = Type.hero),
    headlineMedium = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 18.sp),
    headlineSmall = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 16.sp),
    titleLarge = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = Type.sheet),
    titleMedium = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = Type.heading),
    titleSmall = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = Type.title),
    bodyLarge = TextStyle(fontWeight = FontWeight.Normal, fontSize = Type.title),
    bodyMedium = TextStyle(fontWeight = FontWeight.Normal, fontSize = Type.bodyStrong),
    bodySmall = TextStyle(fontWeight = FontWeight.Normal, fontSize = Type.body),
    labelLarge = TextStyle(fontWeight = FontWeight.Medium, fontSize = Type.bodyStrong),
    labelMedium = TextStyle(fontWeight = FontWeight.Medium, fontSize = Type.body),
    labelSmall = TextStyle(fontWeight = FontWeight.Medium, fontSize = Type.caption)
)

/**
 * The colour scheme is the desktop palette, so a Material component that is
 * not explicitly styled still lands on the right surface — an unstyled
 * `Card` or `DropdownMenu` is the usual way a screen drifts.
 */
private val JXCodeColors = darkColorScheme(
    primary = Palette.accent,
    onPrimary = Palette.accentOn,
    primaryContainer = Palette.accentFill,
    onPrimaryContainer = Palette.accentOn,
    secondary = Palette.tileBlue.fill,
    onSecondary = Palette.tileBlue.ink,
    background = Palette.surfaceDeepest,
    onBackground = Palette.textPrimary,
    surface = Palette.surface,
    onSurface = Palette.textPrimary,
    surfaceVariant = Palette.surfaceElevated,
    onSurfaceVariant = Palette.textSecondary,
    surfaceContainer = Palette.surface,
    surfaceContainerHighest = Palette.surfaceElevated,
    outline = Palette.border,
    outlineVariant = Palette.borderStrong,
    error = Palette.danger,
    onError = Palette.accentOn
)

@Composable
fun JXCodeTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = JXCodeColors,
        typography = JXCodeTypography,
        content = content
    )
}

File: Sources/JXCodeApp/Theme.swift
import JXCodeCore
import SwiftUI

/// The visual language: one design in two modes.
///
/// The references are **one design in two modes** — the same warm neutral base,
/// the same amber accent, the same six dot colours — so this is one palette with
/// two sets of stops rather than two palettes. `Theme` previously pinned a single
/// dark set of constants, which meant a light-mode user got a dark app; the
/// colours here follow the system appearance instead.
///
/// What was sampled, and what it is for:
///
/// - **The base is warm, not blue-grey.** The light page is cream (`#EAE7E1`)
///   under white cards; the dark page is `#141319` under `#191919`. Nothing here
///   is a neutral grey — every surface carries a little red and yellow.
/// - **There is one accent, and it is amber (`#FEB43B`).** It is used as a
///   *fill* with dark text on top, never as coloured text on a light surface,
///   where it would be unreadable. That distinction is why `accent` and
///   `accentFill` are separate: `accent` is the foreground-readable amber (dark
///   in light mode, bright in dark mode) and is what every existing tint call
///   site wants, while `accentFill` is the button colour from the reference.
/// - **The dots are six evenly spaced hues**, sampled from the sidebar: amber,
///   blue, green, teal, purple, violet. They are the same in both modes, because
///   they are identity colours rather than surfaces.
///
/// This file owns the **rendering**; it does not own the values a contrast rule
/// governs. Surfaces, status colours and dots come from `Palette` in
/// `JXCodeCore`, which is where the rule and its tests live — see `Palette` for
/// why the numbers are not here.
enum Theme {

    // MARK: - Surfaces

    /// The page behind everything: the window, the sidebar, the area between
    /// cards. Deepest of the three in both modes.
    static let surfaceDeepest  = adaptive(light: Palette.LightSurface.page,
                                          dark: 0x141319)
    /// Cards and panels — the workhorse surface.
    static let surface         = adaptive(light: Palette.LightSurface.card,
                                          dark: 0x191919)
    /// The selected or hovered fill. Note the direction differs by mode, which
    /// is normal for elevation: in light mode it reads as a pressed tint against
    /// white, in dark mode as a raised panel.
    static let surfaceElevated = adaptive(light: Palette.LightSurface.elevated,
                                          dark: 0x242427)
    /// A well — code, terminal, any area that should read as inset rather than
    /// raised. Dark in **both** modes, which is deliberate and is why the
    /// terminal does not follow the appearance either.
    static let surfaceSunken   = adaptive(light: 0x1F1E1C, dark: 0x0F0E13)

    // MARK: - Lines

    /// Kept as low-alpha inks so they read as a hint, not a stroke, and so the
    /// same constant works on every surface in its mode.
    static let border       = adaptive(light: 0x1F1E1C, dark: 0xFFFFFF,
                                       lightAlpha: 0.10, darkAlpha: 0.08)
    static let borderStrong = adaptive(light: 0x1F1E1C, dark: 0xFFFFFF,
                                       lightAlpha: 0.22, darkAlpha: 0.15)

    // MARK: - Text

    static let textPrimary   = adaptive(light: Palette.TextLight.primary,
                                        dark: Palette.TextDark.primary)
    static let textSecondary = adaptive(light: Palette.TextLight.secondary,
                                        dark: Palette.TextDark.secondary)
    static let textTertiary  = adaptive(light: Palette.TextLight.tertiary,
                                        dark: Palette.TextDark.tertiary)

    // MARK: - Accent

    /// The amber, as a **foreground**: dark enough to read on a light surface,
    /// bright enough to read on a dark one. This is what icon tints and
    /// `.foregroundStyle` want, and it is deliberately not the reference's
    /// button colour.
    static let accent      = adaptive(light: 0x805100, dark: 0xFEB43B)
    /// The amber as a **fill** — the reference's send button and pill. Always
    /// pair with `accentOn`; amber with white text fails contrast in both modes.
    static let accentFill  = adaptive(light: 0xFEB43B, dark: 0xFEB43B)
    static let accentHover = adaptive(light: 0xE89F2C, dark: 0xFFC55E)
    /// Text and glyphs drawn *on* `accentFill`. Also the dark ink for a tile.
    static let accentOn    = adaptive(light: Palette.inkDark, dark: Palette.inkDark)
    // MARK: - Chart segments

    /// Segments of a proportional bar, in draw order.
    ///
    /// These are **not** `tile*` — an identity dot carries its own ink and can
    /// be faint, a bar segment has only the surface behind it and cannot. The
    /// `Palette.Chart*` values are pinned to clear the graphic floor at the
    /// alpha the bar actually draws them at, and to stay far enough apart in
    /// luminance that the bar still reads without hue.
    static let chartWeights   = adaptive(light: Palette.ChartLight.weights,
                                         dark:  Palette.ChartDark.weights)
    static let chartKVCache   = adaptive(light: Palette.ChartLight.kvCache,
                                         dark:  Palette.ChartDark.kvCache)
    static let chartProjector = adaptive(light: Palette.ChartLight.projector,
                                         dark:  Palette.ChartDark.projector)
    static let chartCompute   = adaptive(light: Palette.ChartLight.compute,
                                         dark:  Palette.ChartDark.compute)

    // MARK: - Status

    /// Light values come from `Palette.StatusLight`, where each one is pinned to
    /// clear 4.5:1 on every light surface *and* on its own wash. The dark values
    /// already passed and are kept as sampled.
    static let success = adaptive(light: Palette.StatusLight.success, dark: 0x4BAC65)
    static let warning = adaptive(light: Palette.StatusLight.warning, dark: 0xF5B549)
    static let danger  = adaptive(light: Palette.StatusLight.danger,  dark: 0xF45F59)
    static let info    = adaptive(light: Palette.StatusLight.info,    dark: 0x40C8E0)

    // MARK: - Dots

    static let tileAmber  = solid(Palette.Tile.amber)
    static let tileBlue   = solid(Palette.Tile.blue)
    static let tileGreen  = solid(Palette.Tile.green)
    static let tileTeal   = solid(Palette.Tile.teal)
    static let tilePurple = solid(Palette.Tile.purple)
    static let tilePink   = solid(Palette.Tile.pink)
    static let tileOrange = solid(Palette.Tile.orange)

    /// Deterministic accent for a key. The same workspace always gets the same
    /// dot color so the eye learns where it lives.
    ///
    /// `hashValue` was the obvious choice here and it was wrong: Swift seeds
    /// `String.hashValue` per process, so every relaunch reshuffled the dots
    /// and the promise in the line above held only within a single run. FNV-1a
    /// is stable across processes, which is the entire requirement.
    static func tile(for key: String) -> Color {
        switch stableHash(key) % 6 {
        case 0: return tilePurple
        case 1: return tileOrange
        case 2: return tileTeal
        case 3: return tilePink
        case 4: return tileGreen
        default: return tileAmber
        }
    }

    /// Ink for a glyph drawn on `fill`.
    ///
    /// The rule lives in `Palette.ink(on:)`; this is only the bridge from a
    /// `Color` back to the `0xRRGGBB` it was built from, so there is exactly one
    /// implementation of the decision. A colour that cannot be converted falls
    /// back to white, which is what every tile used to be.
    ///
    /// The ramp is hashed, so any key can land on any dot and every dot has to
    /// carry a readable glyph — white is 1.81:1 on `tileAmber`, which is the
    /// whole reason this is not a constant.
    static func ink(on fill: Color) -> Color {
        guard let rgb = fill.rgbHex else { return .white }
        return solid(Palette.ink(on: rgb))
    }

    /// FNV-1a, 64-bit. Chosen for being short enough to read and stable enough
    /// to depend on — not for distribution quality.
    static func stableHash(_ value: String) -> UInt64 {
        var hash: UInt64 = 0xcbf2_9ce4_8422_2325
        for byte in value.utf8 {
            hash ^= UInt64(byte)
            hash = hash &* 0x0000_0100_0000_01b3
        }
        return hash
    }
}

// MARK: - Building the colours

/// `0xRRGGBB`, sRGB.
extension NSColor {
    convenience init(rgb: UInt32, alpha: Double = 1) {
        self.init(
            srgbRed: CGFloat((rgb >> 16) & 0xFF) / 255,
            green:   CGFloat((rgb >> 8) & 0xFF) / 255,
            blue:    CGFloat(rgb & 0xFF) / 255,
            alpha:   CGFloat(alpha)
        )
    }
}

extension Color {
    /// The `0xRRGGBB` this colour was built from, in sRGB.
    ///
    /// `nil` when the colour cannot be converted to sRGB, which is the honest
    /// answer for a dynamic colour resolved outside an appearance context. Only
    /// the identity dots are read back, and those are the same in both modes.
    var rgbHex: UInt32? {
        guard let srgb = NSColor(self).usingColorSpace(.sRGB) else { return nil }
        func channel(_ value: CGFloat) -> UInt32 {
            UInt32((Double(value) * 255).rounded())
        }
        return (channel(srgb.redComponent) << 16)
             | (channel(srgb.greenComponent) << 8)
             | channel(srgb.blueComponent)
    }
}

/// A colour that resolves per appearance.
///
/// `NSColor(name:dynamicProvider:)` is the mechanism rather than
/// `Color(light:dark:)`-style asset colours because the palette is defined in
/// code, next to the comments that explain where each value came from. A
/// dynamic provider keeps that true and still follows the system appearance.
private func adaptive(light: UInt32, dark: UInt32,
                      lightAlpha: Double = 1, darkAlpha: Double = 1) -> Color {
    Color(nsColor: NSColor(name: nil) { appearance in
        let isDark = appearance.bestMatch(from: [.aqua, .darkAqua]) == .darkAqua
        return isDark
            ? NSColor(rgb: dark, alpha: darkAlpha)
            : NSColor(rgb: light, alpha: lightAlpha)
    })
}

/// A colour that is the same in both appearances.
private func solid(_ rgb: UInt32) -> Color {
    Color(nsColor: NSColor(rgb: rgb))
}

// MARK: - Building blocks

/// A rounded card with a subtle border — the workhorse surface in the sidebar
/// and the action grid. Defaults match `Theme.surface` + `Theme.border`.
struct Card: ViewModifier {
    var padding: CGFloat = 14
    var radius: CGFloat = 10
    var background: Color = Theme.surface
    var border: Color = Theme.border
    func body(content: Content) -> some View {
        content
            .padding(padding)
            .background(
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .fill(background)
            )
            .overlay(
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .stroke(border, lineWidth: 1)
            )
    }
}

extension View {
    func card(padding: CGFloat = 14, radius: CGFloat = 10,
              background: Color = Theme.surface,
              border: Color = Theme.border) -> some View {
        modifier(Card(padding: padding, radius: radius,
                      background: background, border: border))
    }
}

/// Pill-shaped badge for the count chips and the "Free" indicator in the
/// reference. Background defaults to the elevated surface so it reads against
/// both the sidebar and the cards.
struct Badge: View {
    let text: String
    var color: Color = Theme.textSecondary
    var background: Color = Theme.surfaceElevated
    var body: some View {
        Text(text)
            .font(.system(size: 10, weight: .medium))
            .foregroundStyle(color)
            .padding(.horizontal, 7)
            .padding(.vertical, 2)
            .background(Capsule().fill(background))
    }
}

/// Icon tile — a small square with a glyph over a colored background.
/// Used by the workspace list and any future action grid.
///
/// The glyph's ink is **derived from the fill** rather than fixed at white, so a
/// tile stays legible whichever dot the hash picks — see `Theme.ink(on:)`.
///
/// The glyph is an `MXIconName` rather than an SF Symbol string: the app's own
/// chrome is drawn from the vendored mx-icons set, so a typo is a compile error
/// rather than a blank square at runtime.
struct IconTile: View {
    let icon: MXIconName
    var color: Color
    var size: CGFloat = 32
    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 8, style: .continuous)
                .fill(color)
            MXIconView(name: icon, size: size * 0.5, tint: Theme.ink(on: color))
        }
        .frame(width: size, height: size)
    }
}

/// A dot of an identity colour — the reference's sidebar markers, and the
/// smallest unit of the tile palette.
struct ColorDot: View {
    var color: Color
    var size: CGFloat = 8
    var body: some View {
        Circle()
            .fill(color)
            .frame(width: size, height: size)
    }
}

/// The reference's amber primary button: an amber fill with dark text on it.
///
/// Deliberately not `.borderedProminent` with a `.tint`. That style paints its
/// label white, and white on amber fails contrast in both modes — the reference
/// pairs the amber with near-black for exactly this reason. Owning the label
/// colour is the whole reason this exists rather than being a one-line tint.
struct PrimaryButtonStyle: ButtonStyle {
    var compact = false

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .foregroundStyle(Theme.accentOn)
            .padding(.horizontal, compact ? 9 : 12)
            .padding(.vertical, compact ? 3 : 5)
            .background(
                RoundedRectangle(cornerRadius: 7, style: .continuous)
                    .fill(configuration.isPressed ? Theme.accentHover : Theme.accentFill)
            )
    }
}

extension ButtonStyle where Self == PrimaryButtonStyle {
    static var primary: PrimaryButtonStyle { PrimaryButtonStyle() }
    static var primaryCompact: PrimaryButtonStyle { PrimaryButtonStyle(compact: true) }
}
